rectools/models/sasrec.py

```diff
@@ -589,14 +595,21 @@ def _collate_fn_train(
             x[i, -len(ses) + 1 :] = ses[:-1]  # ses: [session_len] -> x[i]: [session_max_len]
             y[i, -len(ses) + 1 :] = ses[1:]  # ses: [session_len] -> y[i]: [session_max_len]
             yw[i, -len(ses) + 1 :] = ses_weights[1:]  # ses_weights: [session_len] -> yw[i]: [session_max_len]
-        return torch.LongTensor(x), torch.LongTensor(y), torch.FloatTensor(yw)
+
+        batch_dict = dict({"x": torch.LongTensor(x), "y": torch.LongTensor(y), "yw": torch.FloatTensor(yw)})
+        if n_negative is not None:
+            negatives = torch.randint(
+                low=1, high=self.item_id_map.size, size=(batch_size, self.session_max_len, self.n_negative)
+            )  # [batch_size, session_max_len, n_negative]
+            batch_dict["negatives"] = torch.LongTensor(negatives)
+        return batch_dict
 
     def get_dataloader_train(self, processed_dataset: Dataset) -> DataLoader:
         """TODO"""
         sequence_dataset = SequenceDataset.from_interactions(processed_dataset.interactions.df)
         train_dataloader = DataLoader(
             sequence_dataset,
-            collate_fn=self._collate_fn_train,
+            collate_fn=lambda batch: self._collate_fn_train(batch, self.n_negative),
```

Comment:
Why can't we access self.n_negative in _collate_fn_train?

Reply:
fixed